backend/src/index.ts

```diff
@@ -159,6 +167,142 @@ function startStatsAggregation(): void {
   setTimeout(regularStatsAggregate, 0);
 }
 
+const updateValidatorDescriptions = async (
+  map: Map<string, ValidatorDescription>
+): Promise<void> => {
+  let currentIndex = 0;
+  while (true) {
+    const metadataInfo = await callViewMethod<PoolMetadataInfo>(
+      "name.near",
+      "get_all_fields",
+      {
+        from_index: currentIndex,
+        limit: VALIDATOR_DESCRIPTION_QUERY_AMOUNT,
+      }
+    );
+    const entries = Object.entries(metadataInfo);
+    if (entries.length === 0) {
+      return;
+    }
+    for (const [accountId, poolMetadataInfo] of entries) {
+      map.set(accountId, {
+        country: poolMetadataInfo.country,
+        countryCode: poolMetadataInfo.country_code,
+        description: poolMetadataInfo.description,
+        discord: poolMetadataInfo.discord,
+        email: poolMetadataInfo.email,
+        twitter: poolMetadataInfo.twitter,
+        url: poolMetadataInfo.url,
+      });
+    }
+    currentIndex += VALIDATOR_DESCRIPTION_QUERY_AMOUNT;
+  }
+};
+
+const getValidatorContractBalance = async (
+  id: string
+): Promise<string | undefined> => {
+  // for some accounts on 'testnet' we can't get 'currentStake'
+  // because they looks like pool accounts but they are not so
+  // that's why we catch this error to avoid unnecessary errors in console
+  return callViewMethod<string>(id, "get_total_staked_balance", {}).catch(
+    () => undefined
+  );
+};
+
+const getPoolInfo = async (id: string): Promise<ValidatorPoolInfo> => {
+  const account = await sendJsonRpcQuery("view_account", {
+    account_id: id,
+    finality: "final",
+  });
+
+  // 'code_hash' === 11111111111111111111111111111111 is when the validator
+  // does not have a staking-pool contract on it (common on testnet)
+  if (account.code_hash === "11111111111111111111111111111111") {
+    return {
+      fee: null,
+      delegatorsCount: null,
+    };
+  }
+  return {
+    // for some accounts on 'testnet' we can't get 'fee' and 'delegatorsCount'
+    // because they looks like pool accounts but they are not so
+    // that's why we catch this error to avoid unnecessary errors in console
+    fee: await callViewMethod<ValidatorPoolInfo["fee"]>(
+      id,
+      "get_reward_fee_fraction",
+      {}
+    ).catch(() => null),
+    delegatorsCount: await callViewMethod<ValidatorPoolInfo["delegatorsCount"]>(
+      id,
+      "get_number_of_accounts",
+      {}
+    ).catch(() => null),
+  };
+};
+
+const getRegularlyFetchedMap = async <T>(
```

Comment:
That's what should be reviewed thoroughly.

We have three maps here:
- map with actual results (for example, for every validator id - its `ValidatorPoolInfo`) (`value map`)
- map with promises of these results (`promise map`)
- map with last called timestamp (`timestamp map`)

This function does following for every id requested (if value does not exist in `promise map` yet):
- adds a promise to `promise map` that updates `value map` as data is fetched
- adds a interval-called function to update promise in `promise map`
- ensures `value map` delete the value if promise fails
- ensures `promise map` delete the value if id was not called for some time (its timestamp in `timestamp map` is older than some value)

Eventually, function resolved when all fetch promises got settled.

This is quite complex, but what we can achieve with this:
- we always have up-to-date data with intervals
- on the first load we can bail out before all the data is fetched and still have partial data (because `value map` is filled as the data gets to the place)

Possible problems:
- fetch all the functions at once may clog RPC of whatever transportation we use
- there may be some edge cases with failures I didn't find
- there may be some race conditions I didn't find